### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Matrix.ToLin

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

- [x] depends on: #3668
- [x] depends on: #3630

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
I fixed the errors, but left unhappy porting notes. If someone else would like to either merge as is, or decide we need to diagnose further, that would be great.

**Comment by jcommelin** [comment]
@semorrison did you forget to push?

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 1]
```diff
@@ -0,0 +1,980 @@
...
+namespace Algebra
+
+section Lmul
+
+variable {R S : Type _} [CommRing R] [Ring S] [Algebra R S]
+
+variable {m : Type _} [Fintype m] [DecidableEq m] (b : Basis m R S)
+
+theorem toMatrix_lmul' (x : S) (i j) :
+    LinearMap.toMatrix b b (LinearMap.Algebra.lmul R S x) i j = b.repr (x * b j) i := by
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/ToLin.lean:844]: `LinearMap.Algebra.lmul` shouldn't exist; #3630 fixes the name

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3630
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3630~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/ToLin.lean:844]: @Ruben-VandeVelde; we'll need to rebase again after the next (two-hourly?) mathport run in order to get this fix.

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 2]
```diff
@@ -0,0 +1,1001 @@
...
+theorem Matrix.toLin'_toMatrix' (f : (n → R) →ₗ[R] m → R) :
+    Matrix.toLin' (LinearMap.toMatrix' f) = f :=
+  Matrix.toLin'.apply_symm_apply f
+#align matrix.to_lin'_to_matrix' Matrix.toLin'_toMatrix'
+
+@[simp]
+theorem LinearMap.toMatrix'_apply (f : (n → R) →ₗ[R] m → R) (i j) :
+    LinearMap.toMatrix' f i j = f (fun j' => if j' = j then 1 else 0) i := by
+  simp only [LinearMap.toMatrix', LinearEquiv.coe_mk, of_apply]
+  congr
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/ToLin.lean:344]: This congr isn't working and I have no idea why. The goal here is
```
↑f (↑(stdBasis R (fun x ↦ R) j) 1) i = ↑f (fun j' ↦ if j' = j then 1 else 0) i
```
`refine FunLike.congr_arg f ?_` doesn't work either

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 3]
```diff
@@ -0,0 +1,1002 @@
...
+    Matrix.toLin' (LinearMap.toMatrix' f) = f :=
+  Matrix.toLin'.apply_symm_apply f
+#align matrix.to_lin'_to_matrix' Matrix.toLin'_toMatrix'
+
+@[simp]
+theorem LinearMap.toMatrix'_apply (f : (n → R) →ₗ[R] m → R) (i j) :
+    LinearMap.toMatrix' f i j = f (fun j' => if j' = j then 1 else 0) i := by
+  simp only [LinearMap.toMatrix', LinearEquiv.coe_mk, of_apply]
+  refine congr_fun ?_ _  -- porting note: `congr` didn't do this
+  congr
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/ToLin.lean:345]: This is a pretty embarrassing failure of `congr`

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3668~~
* ~~leanprover-community/mathlib4#3630~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by eric-wieser** [comment]
@semorrison, the messy merge happened because @Ruben-VandeVelde did a rebase. You probably should have thrown out your local copy and used the version here to make the final fixes, as now the history is a real mess

**Comment by semorrison** [comment]
Please don't force push on public branches!

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**